LUCENE-9965: Add tooling to introspect query execution time

jdconrad opened the pull request:
# Description

Based on the discussion from [this email thread](https://lists.apache.org/thread.html/r7957a2d9ca38af45b1c370753b3c10542fd9faaf9bf95944c5224e12%40%3Cdev.lucene.apache.org%3E) we could add a set of classes to compile timings for different pieces of a query or multiple queries. This could be used to better debug changes in performance moving forward.

# Solution

This change adds a multitude of new classes to help profile query timings. These classes have all been added to the Lucene sandbox including a simple extension of a IndexSearcher with the name ProfileIndexSearcher that includes a QueryProfiler. The QueryProfiler includes the total time spent in each of the following categories along with the number of times visited:

* create weight
* build scorer
* next doc
* advance
* score
* match

# Tests

Tests have been added to ensure that a simple query generate values within a ProfileIndexSearcher.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/lucene/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Lucene maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.

jdconrad committed:
add a set of classes to profile query performance

jdconrad committed:
clean up JavaDoc

jdconrad committed:
add package info

jdconrad committed:
Merge branch 'main' into profile

jtibshirani commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/queries/profile/AbstractInternalProfileTree.java:
```diff
@@ -0,0 +1,185 @@
+ * This class tracks a Query tree for profiling. This class can be extended to allow different
+ * element types for timing with the tree.
+ */
+public abstract class AbstractInternalProfileTree<PB extends AbstractProfileBreakdown<?>, E> {
```
+1 to simplifying the class hierarchy as much as possible. That may mean Elasticsearch can't reuse these classes instead of the original ones, but that seems okay! We could also remove the name `Internal` from these classes to simplify (so we just have `QueryProfileTree` and `ProfileCollector`).

jdconrad replied:
Done. I was thinking it may make sense to preserve the original, but that wouldn't make sense for Lucene.

jtibshirani commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/queries/profile/ProfileIndexSearcher.java:
```diff
@@ -0,0 +1,79 @@
+    super(reader);
+  }
+
+  public void setProfiler(QueryProfiler profiler) {
```
Maybe we could add `QueryProfiler` as a constructor parameter and ensure it's always non-null. That'd simplify this class a bit, and make it clear that it's always meant to be used for profiling.

We could even take this further and remove the `QueryProfiler` class, folding its logic into `ProfileIndexSearcher`. It doesn't seem like a helpful abstraction on its own?

jdconrad replied:
I like this idea a lot as fewer classes in this case is ideal. My question here is do want to have to create a new IndexSearcher for each query?

jtibshirani replied:
I don't have a strong intuition here -- maybe @jpountz has an opinion?

jpountz replied:
Yes, it's totally fine. IndexReaders are costly to create, but IndexSearchers are very cheap.

jtibshirani commented on lucene/sandbox/src/test/org/apache/lucene/sandbox/queries/profile/TestProfileQuery.java:
```diff
@@ -0,0 +1,217 @@
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+
+public class TestProfileQuery extends LuceneTestCase {
```
Small comment, `TestProfileIndexSearcher` could be a clearer name.

jdconrad replied:
Renamed.

jtibshirani commented on lucene/sandbox/src/test/org/apache/lucene/sandbox/queries/profile/TestProfileQuery.java:
```diff
@@ -0,0 +1,217 @@
+    MatcherAssert.assertThat(rewriteTime, greaterThan(0L));
+  }
+
+  public void testCollector() throws IOException {
```
It'd be nice to have one test showing how the collector would actually be used in a search, through something like `IndexSearcher#search(Query query, Collector results)`.

jtibshirani reviewed (commented):
This looks like a good direction to me, just added a few suggestions.

jdconrad commented:
@jpountz @jtibshirani Thank you for the feedback! I will address these soon.

jdconrad committed:
Merge branch 'main' into profile

jdconrad commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/queries/profile/ProfileResult.java:
```diff
@@ -0,0 +1,91 @@
+  }
+
+  /** Retrieve the lucene description of this query (e.g. the "explain" text) */
+  public String getLuceneDescription() {
```
Changed.

jdconrad commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/queries/profile/AbstractProfileBreakdown.java:
```diff
@@ -0,0 +1,76 @@
+  /** Fetch extra debugging information. */
+  protected Map<String, Object> toDebugMap() {
+    return emptyMap();
+  }
```
Removed.

jdconrad commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/queries/profile/AbstractProfileBreakdown.java:
```diff
@@ -0,0 +1,76 @@
+    return emptyMap();
+  }
+
+  public final long toNodeTime() {
```
Changed.

jdconrad commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/queries/profile/AbstractInternalProfileTree.java:
```diff
@@ -0,0 +1,185 @@
+ * limitations under the License.
+ */
+
+package org.apache.lucene.sandbox.queries.profile;
```
Moved.

jdconrad commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/queries/profile/AbstractProfiler.java:
```diff
@@ -0,0 +1,51 @@
+ * This class acts as as storage for a profile tree. This class may be extended to define how the
+ * profile contains and how it's broken into different pieces.
+ */
+public class AbstractProfiler<PB extends AbstractProfileBreakdown<?>, E> {
```
Done. I was thinking it may make sense to preserve the original, but that wouldn't make sense for Lucene.

jdconrad committed:
removed several unnecessary abstractions

jdconrad commented:
Just a note that these comments are still valid and not really outdated. I changed packages and renamed some of the files so it's no longer referencing the correct path.

jtibshirani committed:
Adjust class visibility

jtibshirani committed:
More clean-up

jtibshirani committed:
Fold QueryProfiler into QueryProfilerIndexSearcher

jtibshirani committed:
Refactor collector profiler

jtibshirani committed:
Fix formatting

jpountz commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/search/QueryProfiler.java:
```diff
@@ -0,0 +1,94 @@
+ * QueryProfilerIndexSearcher}. This allows us to know the relationship between nodes in tree
+ * without explicitly walking the tree or pre-wrapping everything.
+ */
+public class QueryProfiler {
```
nit
```suggestion
public final class QueryProfiler {
```

jpountz replied:
Also I believe that all methods but `getTree` and `getCollector` are only for internal usage, should we make them pkg-private?

jpountz commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/search/QueryProfilerBreakdown.java:
```diff
@@ -0,0 +1,66 @@
+ * A record of timings for the various operations that may happen during query execution. A node's
+ * time may be composed of several internal attributes (rewriting, weighting, scoring, etc).
+ */
+public class QueryProfilerBreakdown {
```
This looks like an internal class that could be made pkg-private (and final)?

jpountz commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/search/QueryProfilerCollector.java:
```diff
@@ -0,0 +1,91 @@
+import org.apache.lucene.search.ScoreMode;
+
+/** A collector that profiles how much time is spent calling it. */
+public class QueryProfilerCollector extends FilterCollector {
```
make pkg-private?

jpountz commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/search/QueryProfilerCollectorResult.java:
```diff
@@ -0,0 +1,67 @@
+  protected final String reason;
+
+  /** The total elapsed time for this Collector */
+  protected final Long time;
```
make it a native long?

jpountz commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/search/QueryProfilerCollectorResult.java:
```diff
@@ -0,0 +1,67 @@
+  protected List<QueryProfilerCollectorResult> children;
+
+  public QueryProfilerCollectorResult(
+      String collectorName, String reason, Long time, List<QueryProfilerCollectorResult> children) {
```
use `long` instead of `Long`?

jpountz commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/search/QueryProfilerCollectorWrapper.java:
```diff
@@ -0,0 +1,105 @@
+ *
+ * <p>QueryProfiler facilitates the linking of the Collector graph
+ */
+public class QueryProfilerCollectorWrapper implements Collector {
```
this one looks like it could be pkg-private too?

jpountz commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/search/QueryProfilerResult.java:
```diff
@@ -0,0 +1,83 @@
+ * representation, rather than the entity that collects the timing profile (see InternalProfiler for
+ * that)
+ *
+ * <p>Each InternalProfileResult has a List of InternalProfileResults, which will contain "children"
```
Update to the new names of these classes?

jpountz commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/search/QueryProfilerScorer.java:
```diff
@@ -0,0 +1,208 @@
+ * {@link Scorer} wrapper that will compute how much time is spent on moving the iterator,
+ * confirming matches and computing scores.
+ */
+public class QueryProfilerScorer extends Scorer {
```
can it be pkg-private?

jpountz commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/search/QueryProfilerTimer.java:
```diff
@@ -0,0 +1,97 @@
+ *  }
+ *  </pre>
+ */
+public class QueryProfilerTimer {
```
can it be pkg-private?

jpountz commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/search/QueryProfilerWeight.java:
```diff
@@ -0,0 +1,114 @@
+ * Weight wrapper that will compute how much time it takes to build the {@link Scorer} and then
+ * return a {@link Scorer} that is wrapped in order to compute timings as well.
+ */
+public class QueryProfilerWeight extends Weight {
```
make it pkg-private?

jpountz reviewed (commented):
I left some minor comments around visibility but otherwise this looks good to me!

jtibshirani committed:
Merge remote-tracking branch 'upstream/main' into profile

jtibshirani committed:
Add changelog entry.

jtibshirani commented:
@jdconrad asked that I finish off this work as he'll be away for several weeks. I pushed a few changes:
* Adjust class visibility and other small clean-ups
* Fold QueryProfiler into QueryProfilerIndexSearcher
* Rename profiler collectors and create dedicated test

I'll plan to merge soon unless there are more comments.
